text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,659 @@
+
+After this OTEP is merged, the changes it proposes will be merged into the
+messaging semantic conventions, which will subsequently be declared
+[stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable).
+
+## Proposed stable messaging semantic conventions for tracing
+
+### Terminology
+
+The terminology used in this document is based on the [CloudEvents specification](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md).
+CloudEvents is hosted by the CNCF and provides a specification for describing
+event data in common formats to provide interoperability across services,
+platforms and systems.
+
+#### Message
+
+A "message" is a transport envelope for the transfer of information. The
+information is a combination of a payload and metadata. Metadata can be
+directed at consumers or intermediaries on the message path. Messages are
+transferred via one or more intermediaries.  Messages are uniquely
+identifiable.
+
+In the strict sense, a _message_ is a payload that is sent to a specific
+destination, whereas an _event_ is a signal emitted by a component upon
+reaching a given state. This document is agnostic of those differences and uses
+the term "message" in a wider sense to cover both concepts.
+
+#### Producer
+
+The "producer" is a specific instance, process or device that creates and
+publishes a message. "Publishing" is the process of sending a message or batch
+to the intermediary.
+
+#### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new messages.
+
+The consumer receives, processes, and settles a message.
+
+* "Receiving" is the process of obtaining a message from the intermediary.
+* "Processing" is the process of acting on the information a message contains.
+* "Settling" is the process where intermediary and consumer agree on the state
+  of the transfer.
+
+#### Intermediary
+
+An "intermediary" receives a message for the purpose of forwarding it to the
+next receiver, which might be another intermediary or a consumer.
+
+### Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```
+                               CONSUMER
+
+                   . . . . . . Settle
+PRODUCER           .              ^
+                   .              |
+Create             .           Process
+  |                v              ^
+  v        +--------------+       |
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
```

Comment:
Would you say removing this (fire-and-forget) makes sense then? I can only speculate, but I think the intention was to say that some consumers may chose not to do it, thus being more of a "receive and forget"? I guess @pyohannes would be the ultimate to know the intention for this.

Reply:
I am not sure that "receive and forget" makes sense as well.
In a traditional messaging system, if a consumer gets the message and then doesn't settle it (so it "forgets"), the message will be still available to other consumers at some point after a timeout so it will drive to a duplicate processing of the same message which would be caused not by a network problem (settlement lost for example) but by the consumer intention to "forget". I would say that in general it doesn't make sense, the consumer should settle to allow the system to remove the message from the queue at some point.
In Kafka instead things are different, because the messaging system doesn't have to remove the message from the "topic partition", it will be there for the configured retention period. The difference is in the client which tracks (and commits) the offset of the messages it already read to continue from there (having anyway the way to re-read older messages).

Reply:
MQTT offers a QoS level 0, which is [defined as follows](https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels/):
> The minimal QoS level is zero. This service level guarantees a best-effort delivery. There is no guarantee of delivery. The recipient does not acknowledge receipt of the message and the message is not stored and re-transmitted by the sender. QoS level 0 is often called “fire and forget” and provides the same guarantee as the underlying TCP protocol.

Something similar exists for [AMQP](https://www.ibm.com/docs/en/ibm-mq/9.1?topic=applications-message-delivery-reliability-amqp#q123795___qos).

There is the concept of _at most once delivery_, which is possible with several messaging systems and which some people wanted to have covered by these conventions.

Reply:
Yeah you are right, I forgot that in MQTT for example the consumer can subscribe with a QoS different from the one used by the publisher. Of course, AMQP offers the same with settlement.
Anyway I have never heard talking about "fire and forget" or "receive and forget" so I would avoid using them.
The best way to me sounds to be using:

* AT MOST ONCE, which is a QoS 0, (or "receive and forget")
* AT LEAST ONCE, which is a QoS 1
* EXACTLY ONCE, which is a QoS 2